src/compiler/checker.ts

```diff
@@ -2913,6 +2915,22 @@ namespace ts {
             }
         }
 
+        function baseTypeHasUnappliedOuterTypeParameters(type: Type): boolean {
+            const outerTypeParameters = (<InterfaceType>type).outerTypeParameters;
+            const typeArguments = (<TypeReference>type).typeArguments;
+            if (outerTypeParameters) {
+                // an unapplied type parameter is one 
+                // whose argument symbol is still the same as the parameter symbol
+                const numParameters = outerTypeParameters.length;
+                for (let i = 0; i < numParameters; i++) {

```

Comment:
If it's the case that some type parameters might be applied and some might *not* be applied, are there any that are more likely to be applied than others?

In other words, do we partially apply them towards the beginning? If so, then iterating from the back would bail out more quickly if any type parameter is not applied.